### PR TITLE
`ContractOffer`& `ContractAgreement` reference a single asset (closes #296)

### LINKED ISSUE #296: `Contract` and `ContractOffer` should reference exactly one `AssetId`
Currently the `Contract` and `ContractOffer` objects contain multiple `Asset` references. At the time of writing they reference always one `Asset`, as the `ContractOfferService` maps the `ContractDefinition` like this.
Our domain models should reflect this fact, too, until there is a requirement to support multiple `Asset`.

-> `ContractOffer` should only reference one `AssetId`
-> `ContractAgreement` should only reference one `AssetId`

### PR DESCRIPTION
**Pull request by DominikPinsel**
Signed-off-by: Dominik Pinsel <dominik.pinsel@daimler.com>

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

### CODE REVIEW

**Comment by jimmarino** [comment]
I believe ContractOffer should take a single type `Asset`, not an asset id. That's because the provider needs to communicate asset details to a consumer. This is also related to the `Catalog` type. The `Asset` type contained in `ContractOffer` may be sparsely populated, specifically only having an asset id, for certain message sequences such as when a consumer sends a ContractOffer to a provider. 

For `ContractAgreement` I think we should adopt the same behavior. This will allow us to build an attestation mechanism later by having each party sign the SHA hash of the `ContractAgreement` with a private key and send it to the counter-party.

**Comment by DominikPinsel** [comment]
> I believe ContractOffer should take a single type `Asset`, not an asset id. That's because the provider needs to communicate asset details to a consumer. This is also related to the `Catalog` type. The `Asset` type contained in `ContractOffer` may be sparsely populated, specifically only having an asset id, for certain message sequences such as when a consumer sends a ContractOffer to a provider.
> 
> For `ContractAgreement` I think we should adopt the same behavior. This will allow us to build an attestation mechanism later by having each party sign the SHA hash of the `ContractAgreement` with a private key and send it to the counter-party.

I think so too, but when we discussed this last Friday we discussed exactly this and agreed on a single identifier instead of putting the whole asset in there. That's why I put the Id in there.
I will change the AssetId to Asset and notify you when the changes are done.

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

**Comment by jimmarino** [comment]
> > I believe ContractOffer should take a single type `Asset`, not an asset id. That's because the provider needs to communicate asset details to a consumer. This is also related to the `Catalog` type. The `Asset` type contained in `ContractOffer` may be sparsely populated, specifically only having an asset id, for certain message sequences such as when a consumer sends a ContractOffer to a provider.
> > For `ContractAgreement` I think we should adopt the same behavior. This will allow us to build an attestation mechanism later by having each party sign the SHA hash of the `ContractAgreement` with a private key and send it to the counter-party.
> 
> I think so too, but when we discussed this last Friday we discussed exactly this and agreed on a single identifier instead of putting the whole asset in there. That's why I put the Id in there. I will change the AssetId to Asset and notify you when the changes are done.
> 
We don't need to put the entire asset in for all cases, it can be sparsely populated. @paullatzelsperger suggested that.

**Comment by DominikPinsel** [comment]
@jimmarino updated the PR

**Review by jimmarino — APPROVED** [review]
Thanks

**Comment by jimmarino** [comment]
@DominikPinsel can you look at the conflicts and I will then merge?

**Comment by DominikPinsel** [comment]
> @DominikPinsel can you look at the conflicts and I will then merge?

@jimmarino done :)

**Review by jimmarino — APPROVED** [review]

**Comment by paullatzelsperger** [comment]
i fixed the `equals()` and `hashCode()`, cf this [gist](https://gist.github.com/paullatzelsperger/54ffd19ee41621f6eba0b6e94ce3cec5)

**Comment by jimmarino** [comment]
> i fixed the `equals()` and `hashCode()`, cf this [gist](https://gist.github.com/paullatzelsperger/54ffd19ee41621f6eba0b6e94ce3cec5)


@DominikPinsel can you apply this patch to fix Asset equals and hashcode?

**Comment by DominikPinsel** [comment]
> > i fixed the `equals()` and `hashCode()`, cf this [gist](https://gist.github.com/paullatzelsperger/54ffd19ee41621f6eba0b6e94ce3cec5)
> 
> @DominikPinsel can you apply this patch to fix Asset equals and hashcode?

@jimmarino Done
@paullatzelsperger Thanks!

File: data-protocols/ids/ids-api-multipart-dispatcher-v1/src/test/java/org/eclipse/dataspaceconnector/ids/api/multipart/client/MultipartDispatcherIntegrationTest.java [hunk 1]
```diff
@@ -115,8 +115,9 @@ void testSendDescriptionRequestMessage() throws Exception {
     @Test
     void testSendArtifactRequestMessage() throws Exception {
         var asset = (Asset) EasyMock.createNiceMock(Asset.class);
-        EasyMock.expect(asset.getId()).andReturn("assetId");
+        EasyMock.expect(asset.getId()).andReturn("1");
```
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply data-protocols/ids/ids-api-multipart-dispatcher-v1/src/test/java/org/eclipse/dataspaceconnector/ids/api/multipart/client/MultipartDispatcherIntegrationTest.java]: done

File: data-protocols/ids/ids-transform-v1/src/main/java/org/eclipse/dataspaceconnector/ids/transform/IdsResourceCatalogToDataCatalogTransformer.java [hunk 2]
```diff
@@ -113,7 +112,7 @@ private ContractOffer createContractOffer(de.fraunhofer.iais.eis.ContractOffer o
                 .provider(object.getProvider());
 
         if (asset != null) {
```
[the remark this answers is not in the document]
Comment (DominikPinsel) [reply data-protocols/ids/ids-transform-v1/src/main/java/org/eclipse/dataspaceconnector/ids/transform/IdsResourceCatalogToDataCatalogTransformer.java]: done

**Pull request merged**